packages/@expo/server/src/index.ts

```diff
@@ -126,6 +147,54 @@ export function createRequestHandler(
 
     debug('Request', sanitizedPathname);
 
+    // Execute middleware first, before any route matching
+    if (routesManifest.middleware) {
+      try {
+        const middlewareModule = await getMiddleware(routesManifest.middleware);
+        if (middlewareModule?.default) {
+          const middlewareResponse = await middlewareModule.default(request);
+          if (middlewareResponse instanceof Response) {
+            debug('Middleware returned response, short-circuiting');
+            return middlewareResponse;
+          }
+          // If middleware returns undefined/void, continue to route matching
+        }
+      } catch (error) {
+        // Only show detailed errors for development; in production, show a generic error to prevent leaking sensitive information
+        debug('Middleware execution error:', error);
+
+        const isApiRequest = routesManifest.apiRoutes.some((route) =>
+          route.namedRegex.test(sanitizedPathname)
+        );
+
+        if (isApiRequest) {
+          return new Response(
+            JSON.stringify({
+              error:
+                process.env.NODE_ENV === 'development'
+                  ? error
+                  : { message: 'Internal Server Error' },
+            }),
+            {
+              status: 500,
+              headers: { 'Content-Type': 'application/json' },
+            }
+          );
+        } else {
+          if (process.env.NODE_ENV === 'development') {
+            return handleApiRouteError(error as Error);
+          }
+
+          // TODO(@hassankhan): Should we return any HTML at all?
+          const errorHtml = `<!DOCTYPE html><html><head><title>Server Error</title></head><body><h1>Internal Server Error</h1></body></html>`;
```

Comment:
Maybe something like `+error.ts` (similar to `+not-found.ts`) could be handy in this case. 

But as a users I can now `try catch` the middleware code and return my own page, right?

Reply:
Yeah I'd love for us to have a `+error.ts` or similar in the future. For now, I'll defer to @EvanBacon on what to do for error handling in production for both API and HTML routes.